be/src/runtime/stream_load/transaction_mgr.cpp

```diff
@@ -397,30 +397,29 @@ void TransactionMgr::_clean_stream_context() {
         auto ctx = _exec_env->stream_context_mgr()->get(id);
         if (ctx != nullptr) {
             int64_t now = UnixSeconds();
-            // try lock fail means transaction in processing
-            if (ctx->lock.try_lock()) {
-                // abort timeout transaction
-                if ((now - ctx->begin_txn_ts) > ctx->timeout_second && ctx->timeout_second > 0) {
-                    ctx->status = Status::Aborted(fmt::format("transaction is aborted by timeout."));
-                    auto st = _rollback_transaction(ctx);
-                    LOG(INFO) << "Abort transaction " << ctx->brief() << " since timeout " << ctx->timeout_second
-                              << " begin ts " << ctx->begin_txn_ts << " status " << st;
-                }
-
-                if (ctx->body_sink != nullptr) {
-                    if (!ctx->body_sink->exhausted()) {
+            Status status;
+            if ((now - ctx->begin_txn_ts) > ctx->timeout_second && ctx->timeout_second > 0) {
+                status = Status::Aborted(
+                        fmt::format("transaction is aborted by timeout {} seconds.", ctx->timeout_second));
+            } else if ((now - ctx->last_active_ts) > ctx->idle_timeout_sec + interval && ctx->idle_timeout_sec > 0) {
+                status = Status::Aborted(fmt::format("transaction is aborted by idle timeout {} seconds.",
+                                                     ctx->idle_timeout_sec + interval));
+            } else {
+                std::shared_ptr<MessageBodySink> body_sink = ctx->body_sink;
+                if (body_sink != nullptr) {
+                    if (!body_sink->exhausted()) {
                         ctx->last_active_ts = UnixSeconds();
                     }
                 }
-
-                if ((now - ctx->last_active_ts) > ctx->idle_timeout_sec + interval && ctx->idle_timeout_sec > 0) {
-                    ctx->status = Status::Aborted(fmt::format("transaction is aborted by idle timeout."));
+            }
+            if (!status.ok()) {
+                if (ctx->lock.try_lock()) {
+                    ctx->status = status;
```

Comment:
_rollback_transaction() can be a long time op as it may invoke FE rpc. the shorten lock in the above check might not be the main factor contributing to the txn_in_progress?

How about turn to timed_mutex with try_lock_for, so it can wait for a moment if can't lock immediately?

Another point, if the ctx-status can be taken as the exclusive resource between different threads, if the status doesn't match the expected one, just respond error, so the _rollback_transaction might not be necessary to be in the lock protection, or read lock is good enough?

Reply:
`_rollback_transaction` is only called after the load timeouts. However, based on some cases I have analyzed, they all occurred before the timeout, which suggests that `_rollback_transaction` is not the root cause, but rather the timeout checking. I also have verified it in a user's environment.

Reply:
how much time cost for the timeout check without rollback transaction op? a few microseconds or even less than single microsecond? Still doubt if rollback transaction is still under lock, how much will be improved for the conflict.

Nevertheless, try lock with a timeout about milliseconds, will be still worth instead of just try_lock and fail with txn_in_progress error. It is an REST API call, not a function call, a delay of milliseconds is still fast enough and probably get the determined result other than TXN_IN_PROGRESS immediately.

Reply:
1. At least from the existing cases, rollback does not seem to improve the conflict, as the logs show that rollback was not triggered. However, improvements can be made to the error messaging here. If rollback is triggered, the request should still fail, but the error message should be changed to "The load is timeout, and will be aborted". see https://github.com/StarRocks/starrocks/pull/54959/commits/1a8ab3195d4283c2b71e7c0b4c8b626752eebd19
2. no matter which conflicts, no need to wait and should fail the request immediately. The key is giving the right error message as 1 said